src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.PendingBranchesCollection.cs

```diff
@@ -0,0 +1,146 @@
+
+            private void FreeBranchesToLabel()
+            {
+                if (_branchesToLabel is { })
+                {
+                    foreach (var branches in _branchesToLabel.Values)
+                    {
+                        branches.Free();
+                    }
+                    _branchesToLabel.Free();
+                    _branchesToLabel = null;
+                }
+            }
+
+            internal ImmutableArray<PendingBranch> ToImmutable()
+            {
+                return _branchesToLabel is null ?
+                    _branches.ToImmutable() :
+                    ImmutableArray.CreateRange(this);
+            }
+
+            internal ArrayBuilder<PendingBranch> GetAndRemoveBranches(LabelSymbol? label)
+            {
+                ArrayBuilder<PendingBranch>? result;
+                if (label is null)
+                {
+                    result = _branches;
+                    _branches = ArrayBuilder<PendingBranch>.GetInstance();
+                }
+                else if (_branchesToLabel is { } && _branchesToLabel.TryGetValue(label, out result))
+                {
+                    _branchesToLabel.Remove(label);
+                }
+                else
+                {
+                    result = ArrayBuilder<PendingBranch>.GetInstance();
+                }
+                return result;
+            }
+
+            internal void Add(PendingBranch branch)
+            {
+                var label = branch.Label;
+                if (label is null)
+                {
+                    _branches.Add(branch);
+                }
+                else
+                {
+                    var branches = GetOrAddBranchesToLabel(label);
+                    branches.Add(branch);
+                }
+            }
+
+            internal void AddRange(PendingBranchesCollection collection)
+            {
+                _branches.AddRange(collection._branches);
+                if (collection._branchesToLabel is { })
+                {
+                    foreach (var pair in collection._branchesToLabel)
+                    {
+                        var branches = GetOrAddBranchesToLabel(pair.Key);
+                        branches.AddRange(pair.Value);
+                    }
+                }
+            }
+
+            private ArrayBuilder<PendingBranch> GetOrAddBranchesToLabel(LabelSymbol label)
+            {
+                if (_branchesToLabel is null)
+                {
+                    _branchesToLabel = PooledDictionary<LabelSymbol, ArrayBuilder<PendingBranch>>.GetInstance();
+                }
+                if (!_branchesToLabel.TryGetValue(label, out var branches))
+                {
+                    branches = ArrayBuilder<PendingBranch>.GetInstance();
+                    _branchesToLabel.Add(label, branches);
+                }
+                return branches;
+            }
+
+            public IEnumerator<PendingBranch> GetEnumerator()
```

Comment:
The enumerator here is non-deterministically ordered. This seems to fan out into other data structures not being deterministically ordered as well like `ExitPointsWalker._branchesOutOf` and `LocalDataFlowPass.Join` calls. Want to make sure that was accounted for with this change.